lib/ts/recipe/oauth2/types.ts

```diff
@@ -0,0 +1,298 @@
+/* Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ * This software is licensed under the Apache License, Version 2.0 (the
+ * "License") as published by the Apache Software Foundation.
+ *
+ * You may not use this file except in compliance with the License. You may
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { BaseRequest, BaseResponse } from "../../framework";
+import OverrideableBuilder from "supertokens-js-override";
+import { GeneralErrorResponse, JSONObject } from "../../types";
+import { SessionContainer } from "../session";
+
+export type TypeInput = {
+    override?: {
+        functions?: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis?: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeNormalisedInput = {
+    override: {
+        functions: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type APIOptions = {
+    recipeImplementation: RecipeInterface;
+    config: TypeNormalisedInput;
+    recipeId: string;
+    isInServerlessEnv: boolean;
+    req: BaseRequest;
+    res: BaseResponse;
+};
+
+export type APIInterface = {
+    tokenPOST(input: {
+        clientId: string;
+        grantType: GrantType;
+        clientSecret?: string;
+        code?: string;
+        codeVerifier?: string;
+        refreshToken?: string;
+        redirectUri?: string;
+        scope: string[];
```

Comment:
should it be scope or scopes?

Reply:
This maps to a query param called `scope` and I wanted to keep that consistent. Internally, we use plural everywhere because it indicates that it is an array, but I didn't want to make this change on the API level. Although, we are making a change from snake_case to camelCase anyway, so if you have a strong preference I can rename this as well.

Reply:
I think internally, we should also call it `scope` then.